Tasty Reader with Simple Trees in Annotations [ci: last-only]

bishabosha opened the pull request:
This pull request adds the ability for the SymbolTable to be populated from a classfile that is linked to a TASTy file by reading its contents. Similar to how Pickles provide symbols. The TASTy reader is not a tree provider, however there is limited support for reading expressions, enough to read annotation trees with constants and paths. Future work would be to add full support for creating trees for defs, vals, templates etc.

I would greatly appreciate feedback on a few things that I will point out in the diffs, but mainly how I handle adaption of value class extension methods, and dotty's enum constants.

## Steps to Review
i. Be sure to try out the flag` -Ydebug-tasty` which gives a lot of information which you can use to trace where in a tasty file you are currently traversing
1) About half of the lines changed and the majority of files are related to testing
2) It is probably not worth reviewing commit by commit as many designs are refactored or cut along the way
3) Begin within `src/compiler/scala/tools/tasty/` - these classes provide machinery for reading bytes from TASTy, and some simple data structures that represent static data in TASTy:
   - `TastyFormat.scala`, `TastyHeaderUnpickler.scala`, `TastyRefs.scala`, `UnpickleException.scala`, `TastyReader.scala` - these are duplicates of sources from the dotty repo, unchanged except to remove dead code. These purely deal with traversing TASTy bytes
   - `TastyFlags.scala`: A static type representing a bitset of flags that are encoded in a TASTy file, along with some standard flags inferred from context, such as `Method` and `Deferred`.
   - `TastyName.scala`: This is a data structure representing semantic names. `TastyName` is the interface that TASTy uses to select members from a type, providing more information than simple strings, such as selecting a Types over Terms, companion module instead of a class, or if a term is a default getter. Names can also be paired with a `MethodSignature[ErasedTypeRef]`, which is used to select an overloaded method.
   - `ErasedTypeRef.scala`: This is a wrapper around a `TastyName.TypeName` that adds a count of the number of array dimensions the referenced type has. It carries the invariant that it references the erasure of a fully qualified class or module, and never any other definition.
   - `Signature.scala`: Currently only encodes `MethodSignature`, which represents the erased signature of a method, along with any interspersed type parameter lists 
4) Next in `src/compiler/scala/tools/nsc/tasty` are compiler specific definitions for reading tasty:
   - `TastyModes.scala`: A static type representing a bitset of modes that affect the interpretation of a TASTy file, such as distinguishing between reading the parents of a class, or an annotation tree.
   - `package.scala`: provides `SafeEq`, which is a implicit value class adding equality operators asserting at compiletime that the rhs is a subtype of the lhs.
   - `TastyUniverse.scala`: provides a wrapper around `scala.tools.nsc.symbtab.SymbolTable` adapting it to the API TASTy requires to resolve types and symbols and create trees. Uses a cake with dependent types from symbolTable.
   - `ScalacUnpickler.scala`: the entry point to TASTy unpickling for `nsc`, initialises a `TastyUniverse#Context` with the root symbols of a top level class, then delegates to `TastyUnpickler` to parse names from a TASTy file and divide it into sections, before entering symbols from the `ASTs` section with `TreeUnpickler`
   - `TastyUnpickler.scala`: Divides a TASTy file into sections, reading names into a cache of `TastyName`. Responsible for providing a `TastyReader` associated with a section to a `SectionUnpickler`.
   - `TreeUnpickler.scala`: Responsible for traversing all trees in a TASTy file, representing the definitions inside the classfile associated with the root class/module and extracting the public API into the symbolTable. Public API includes annotations where they are simple trees. Where possible, `TreeUnpickler` should not directly manipulate values created by the `symbolTable`, but use operations provided by `TastyUniverse`
5) Next in `src/compiler/scala/tools/nsc/tasty/bridge/` are the sources that make up the `TastyUniverse` cake, divided by responsibility:
   - `TastyCore.scala`: the base of the cake, provides aliases to types from `scala.reflect` at the same import level as new TASTy specific types.
   - `ContextOps.scala`: This is the most important trait, this contains the definition for `Context`, which is used to maintain state through traversal of a TASTy file, such as the current owner - the container of the current definition; the traversal mode; and the root owners and source path for the TASTy file. It also provides operations for manipulation of the symbol table, such as creating/updating symbols and finding fully qualified classes/modules. the file also provides some standard error throwing capabilities with formatted errors that depend on the context mode.
   - `AnnotationOps.scala`: adds support for creating annotations from trees
   - `FlagOps.scala`: Handles encoding of `TastyFlags` to `scala.reflect` flags and witnessing which flags do not map directly from TASTy.
   - `NameOps.scala`: This layer handles encoding of `TastyName` to `symbolTable.Name`, escaping any specially handled names. Also contains definitions of names for handling special dotty internal symbols within `TreeUnpickler`.
   - `SymbolOps.scala`: This layer deals with selecting a member symbol from a type using a `TastyName`, also adds factories for making type references to symbols
   - `TreeOps.scala`: This layer adds factories that construct typed `scala.reflect` Trees in the shapes that TASTy expects
   - `TypeOps.scala` This layer adds factories that construct `scala.reflect` Types in the shapes that TASTy expects. Additionally provides operations to select a type from a prefix, or a type with an additional prefix, using a `TastyName`.
6) Additionally we have some extra changes to internals outside of the new `tasty` packages
   - `src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala`: Adds logic to look for the TASTY attribute in a classfile and then delegate to `ScalacUnpickler`, similar to what is done for `UnPickler`
   - `src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala`: Moved logic for creating extension method symbols to `internal.Types`, as it needs to be triggered from within the `TreeUnpickler`. I believe it would be too inefficient to modify `ExtensionMethods.scala` to inspect all select nodes to check if the symbol owner is a value class. Perhaps there is another efficient way to generate extension methods for value classes from TASTy.
   - `src/compiler/scala/tools/nsc/transform/patmat/PatternMatching.scala`: Added logic that checks the selector of a match if it has the annotation `scala.annotation.internal.Child`, which is the dotty encoding of sealed children. This has potential to be moved to be driven from within `TreeUnpickler` but there is the problem of forcing the annotation at the correct time to prevent errors from reading a forward reference, and forcing each annotation to ensure exhaustivity.
   - ~`src/reflect/scala/reflect/internal/Constants.scala`: Added a new `DottyEnumTag` which is used to give a precise type to a dotty enum constant before erasure. After erasure these should be treated the same as a java enum. If dotty enum constants are not typed with a constant type then pattern match warnings use a wildcard pattern and not the constants name - perhaps pattern match logic could be adjusted rather than adding a new constant kind. This change also affects `BCodeBodyBuilder.scala`, `internal/Definitions.scala` and `internal/Importers.scala`.~ - this change has been reversed.
    - `src/reflect/scala/reflect/internal/transform/Erasure.scala`: added a check that when a dotty enum singleton is erased, it is no longer a module class, but erased like the intersection of its parents. This is matched by a change to `src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala` which is modified to that accesses to Dotty enum singletons are to a field in the wrapping object, not a field within the module class of the symbol
7) Testing - A new testing framework `TastyTest`, has been added in `src/tastytest/scala/tools/tastytest/` which recreates (pos/neg/run) scenarios which the tasty reader is intended to be used for: typically, compiling some Scala 3 code with dotc and using the Dotty library, then depending on it from Scala 2 code. the modes are explained in `doc/internal/tastytest.md`. `TastyTest` also has test kinds to force dependencies compiled with dotc to not be available on the classpath from scala 2. Alternative: Partest could be adopted to integrate these new workflows, or perhaps it already has a mode for dependent compilations using a different compiler between stages.
   - `build.sbt` several task keys have been added to assist with testing individual scenarios in the `tasty_reader`, perhaps these pollute the build?
   - `test/tasty` This directory consists exclusively of test sources to be compiled by `TastyTest`, which is automated by a `JUnit` suite: `test/tasty/test/scala/tools/tastytest/TastyTestJUnit.scala`, which wraps `TastyTest` suites with customised options.

### Dotty Support:
There is an attempt to provide some more information, when possible, about unsupported dotty features, (currently, union types can not be pretty printed, with more work it is possible to flatten nested union types for the purpose of better error messages)

bishabosha committed:
correctly enter package

bishabosha committed:
intercept java.lang.Object parent

bishabosha committed:
Add signed name table; read signed name first

bishabosha committed:
Support for resolving overloaded signatures for methods that are not curried

bishabosha committed:
Support curried overloads with signatures, WIP on HK signatures

bishabosha committed:
rename TASTY -> Tasty; delete unneeded code

bishabosha committed:
Factor out dotc APIs to TastyUniverse

bishabosha committed:
delete more code

bishabosha committed:
support template parents with type parameters

bishabosha committed:
unpickle Functor from TASTY

bishabosha committed:
use typeSymbolDirect

bishabosha committed:
Unpickle bounded variants of Functor

bishabosha committed:
Update README and publish tests

bishabosha committed:
ignore Extension flag, suspend unpickling Monad

bishabosha committed:
Avoid too precise type for TYPEREFsymbol; restore TestMonad

bishabosha committed:
change TastyFlags to AnyVal

bishabosha committed:
TestReader blocked by dotc

bishabosha committed:
unpickle companions in same tasty file; suspend TestPair; replace EmptyProduct with Box

bishabosha committed:
WIP: Inner classes

bishabosha committed:
unpickle inner traits and objects; restore TestInner

bishabosha committed:
unpickle traits with abstract type members

bishabosha committed:
support inheritance of object member

bishabosha committed:
unify handling of hk params on methods and parents

bishabosha committed:
unpickle value class with private ctor

bishabosha committed:
create suspended TestEnum

bishabosha committed:
Restore TestEnum, do not unpickle typeParams for constructors

bishabosha committed:
Add simple regression test runner for tasty files

bishabosha committed:
cleanup TastyTest

bishabosha committed:
test lazy val

bishabosha committed:
Restore TestPair, initial support for module TYPEREFs

bishabosha committed:
create suspended TestExtensions

bishabosha committed:
Document TastyTest framework

bishabosha committed:
create suspended TestJColour for j.l.Enum

bishabosha committed:
remove dependency on dotc and coursier on the path

bishabosha committed:
add -help to TastyTest, invoke scala compiler directly, add args for sources and package name

bishabosha committed:
remove early initialiser in ScalacUnpickler

bishabosha committed:
cleanup TastyTest

bishabosha committed:
make reframe TastyTest as a runner with multiple suite kinds

bishabosha committed:
Support neg tests in TastyTest and add an example neg test

bishabosha committed:
accumulate errors in TastyTest suites

bishabosha committed:
rework whitelisting sources

bishabosha committed:
refactor Runner

bishabosha committed:
Correctly capture thrown errors in run tests, experiment with test suite super class

bishabosha committed:
collective errors on Suite

bishabosha committed:
sync with standalone tastytest version

bishabosha committed:
invoke dotc in same jvm process

bishabosha committed:
Protect against thrown errors in scalac and dotc; migrate all tests to Suite

bishabosha committed:
Add positive super call test; suspend symbollic operator test

bishabosha committed:
update tests and README for dotty 0.20.0-RC1

bishabosha committed:
make sbt projects for tastytest, use new diff tool

bishabosha committed:
add tasty/test to testAll; disable other tests

bishabosha committed:
implement PARAMEND

bishabosha committed:
implement open modifier

bishabosha committed:
correct PARAMS tag value

bishabosha committed:
depend on dotty with tasty version 18

bishabosha committed:
Split up tastytest

bishabosha committed:
Update docs and remove pre from neg tests

bishabosha committed:
add test output

bishabosha committed:
Add Name ADT to scala.tools.nsc.tasty

bishabosha committed:
Reify Tasty Names, which allows to reduce coupling

bishabosha committed:
reduce dependencies on symbolTable

bishabosha committed:
make TastyUniverse a parameter not a parent

bishabosha committed:
unpickle is now extension method

bishabosha committed:
default canonical LambdaType now PolyType, add pos suites

bishabosha committed:
replace def forwarders to objects with forwarders to factory method

bishabosha committed:
downgrade to dotty 0.20.0-RC1

bishabosha committed:
support SHAREDterm, add neg-false test

bishabosha committed:
move enum flag to TastyFlags. neg-false test for scala2 style enum pattern

bishabosha committed:
add different SColour test for run

bishabosha committed:
fix diff tools, prepare IO test

bishabosha committed:
unpickle byname type param, move TestIO to pos

bishabosha committed:
ensure bynametype has function type. Allow to skip check on neg test

bishabosha committed:
better logging output

bishabosha committed:
minimise type intersections issue. Identify dotty intersecion or union types

bishabosha committed:
Resolve intersection types

bishabosha committed:
add Implicit flag not Given

bishabosha committed:
Add nameTagToString to TastyFormat

bishabosha committed:
unpickle inlineaccessor and superaccessor names

bishabosha committed:
improve error messages

bishabosha committed:
adjust scala 3 flags show output

bishabosha committed:
unpickle intersection types in type position

bishabosha committed:
dotty macros now have their own flag, stub factor10 intent library for testing macros

bishabosha committed:
update TODO, add test for trait initialisers

bishabosha committed:
add support for unpickling unescaped operator names; unsuspend TestProd; add TestSetters

bishabosha committed:
can unpickle some backticked names, will have to wait until dotty 0.21.0-RC1 release for all

bishabosha committed:
more precise Intersection test. Assert that default arguments are unrecognised

bishabosha committed:
unpickle overloaded methods

bishabosha committed:
unpickle default parameters

bishabosha committed:
Special case more names

bishabosha committed:
Add regression test for empty package

bishabosha committed:
add to scolour test

bishabosha committed:
Unpickle extension methods

bishabosha committed:
extract extension method symbol and type factorys

bishabosha committed:
move extension method symbol generation

bishabosha committed:
WIP

bishabosha committed:
add pos-false test, resolve reachability issue, must solve exhaustion analysis

bishabosha committed:
Tasty signatures now support array types

bishabosha committed:
Read vararg trees

bishabosha committed:
unpickle annotations, special case scala.annotation.internal.Child for sealed case enumeration

bishabosha committed:
signatures now look for module in mirror, restore ModuleInCtor

bishabosha committed:
unpickle singleton types correctly in arguments

bishabosha committed:
refactor typeref logic

bishabosha committed:
refactor annotation logic

bishabosha committed:
Unpickle If (tested with an annotation)

bishabosha committed:
support NamedArg tree unpickle, begin support for match

bishabosha committed:
avoid caching a LambdaType, put more complex expressions in parents

bishabosha committed:
resolve arbitrary tree in ctor problem

bishabosha committed:
Modify test functor; ensure folders are published

bishabosha committed:
fix #25 - Handle a signed name in a TERMREF

bishabosha committed:
remove todo document, direct users to issues

bishabosha committed:
fix #27: Add tasty-ci

bishabosha committed:
fix breakages after rebase

bishabosha committed:
fix #24: use custom nsc reporter in neg tests

bishabosha committed:
use merge commit instead

bishabosha committed:
update README

bishabosha committed:
try removing the push event

bishabosha committed:
update README

bishabosha committed:
identify crash with test

bishabosha committed:
flush reporter in TastyTest

bishabosha committed:
temporarily store a module symbol to ensure its symbol survives, allow modules to be overloaded

errikos committed:
Add HKNest test to test higher kinded types

Fails with:

TestHKNest.scala:10: error: no type parameters for method foo: (x: tastytest.HKNest.Foo[F])String exist so that it can be applied to arguments (tastytest.HKNest.Foo[tastytest.HKNest.Box[_$8]])
 --- because ---
argument expression's type is not compatible with formal parameter type;
 found   : tastytest.HKNest.Foo[tastytest.HKNest.Box[_$8]]
 required: tastytest.HKNest.Foo[?F]30s

    test(assert(new Consume[Foo].foo(new Foo[Box]) == "Foo"))
                                 ^
TestHKNest.scala:10: error: type mismatch;
 found   : tastytest.HKNest.Foo[tastytest.HKNest.Box[_$8]]
 required: tastytest.HKNest.Foo[F]
    test(assert(new Consume[Foo].foo(new Foo[Box]) == "Foo"))
                                     ^
2 errors

bishabosha committed:
constructor returns type of owner

bishabosha committed:
test consuming (* -> *) -> * from TASTy

bishabosha committed:
inherit typeParams from bounds of hk lambda types; add many more tests

bishabosha committed:
refactor to new mkTypeRef factory; test HK nest with covariant bounds

bishabosha committed:
simplify lambdatype canonical repr, handle contravariance

bishabosha committed:
normalise typebounds in hktypeparams, handle wildcard bounds, pretty print wildcard idents

bishabosha committed:
cleanup

bishabosha committed:
add more test

bishabosha committed:
Add test that causes error when compiled with dotty

bishabosha committed:
WIP, investigate bounds stripping for lower bounds with polytype

bishabosha committed:
normalise typebounds in hk types

bishabosha committed:
refactor

julienrf committed:
Add `@compileTimeOnly` annotation to unpickled inline def definitions

Fixes scalacenter/scala#9.

errikos committed:
Drop package prefix from a class

errikos committed:
Add .keep file to neg-false/src-2

bishabosha committed:
merge parameters of type lambdas in bounds

bishabosha committed:
error on curried type lambda

bishabosha committed:
refactor

bishabosha committed:
refactor

bishabosha committed:
survive more pathological cases of type lambdas in bounds

bishabosha committed:
improve error reporting

bishabosha committed:
refactor LambdaEncoding

bishabosha committed:
refactor lambdaTypeCompanion

bishabosha committed:
refine symbol table settings type in TastyKernel

bishabosha committed:
add test to show flag disables annotations

bishabosha committed:
upgrade to TASTy version 18, refactor flags

bishabosha committed:
add tasty/dotc and tasty/scalac tasks for ease of use

bishabosha committed:
add more sbt tasks: code runner and dotty decompiler

bishabosha committed:
change to Dotty 0.21.0, update readme, simplify test runner

bishabosha committed:
refactor

bishabosha committed:
Upgrade tasty version 19, drop variant names and adjust bounds

bishabosha committed:
clone LambdaPolyType with new variances

bishabosha committed:
revert to in place update of variance flags

bishabosha committed:
support dotty enum constants

bishabosha committed:
Ensure paths are preserved for type member selection

bishabosha committed:
Add test case for refined types

bishabosha committed:
Add passing REFINEDtype run and neg tests

bishabosha committed:
successful tests for REFINEDtpt including compound types

bishabosha committed:
add path dependent type test on dotty enum constant

bishabosha committed:
support val and nullary def in REFINEDtype

bishabosha committed:
Test methodic refinements for REFINEDtpt

bishabosha committed:
Support refinements for method, poly and values, support structural refinements

bishabosha committed:
support repeated method parameters

bishabosha committed:
Update to use repeated parameters

bishabosha committed:
unpickle SUPER and SUPERtype symbols

bishabosha committed:
Support Implicit and Given flags on method refinements

bishabosha committed:
add test cases

sjrd committed:
Fix `tasty/test` on Windows.

There were still two path separator issues.

bishabosha committed:
add more idiomatic structural refinements

bishabosha committed:
Add nested RECtype test case

bishabosha committed:
Tests pass for non-nested RECtype

bishabosha committed:
support forward-referencing RECthis and type alias refinements

bishabosha committed:
Add test for path dependent val refinement

bishabosha committed:
add test cases for TERMREFin/TYPEREFin

bishabosha committed:
Support TERMREFin and TYPEREFin, look for type members in type ctor

bishabosha committed:
Support Dotty 0.23.0-RC1; opaque type aliases are transparent

bishabosha committed:
re-enable tests

bishabosha committed:
add new github action to run on push to tasty_reader

bishabosha committed:
Cleanup warnings

bishabosha committed:
delete dead code and comments, refactor

bishabosha committed:
refactor reading a member by name, enabling dotty java enums

bishabosha committed:
This is a very disruptive change, affecting bytecode gen (could be avoided if I
can time travel to change the info of a dotty enum constant for erasure)

This enables type checking to work for a dotty enum constant with a mixin if
the expected type is a mixin and not the enum parent class

bishabosha committed:
Fail fast when select name from type is no symbol

bishabosha committed:
Add TastyMode, test when parent is missing from classpath

bishabosha committed:
Add test when type in a method is missing from classpath

bishabosha committed:
Add test when annotation on top-level class is missing from classpath

bishabosha committed:
refactor tasty name, add ErasedTypeRef, special case repeated class

bishabosha committed:
unify error handling and logging

bishabosha committed:
remove module symbol cache from context

bishabosha committed:
unify attachTypeError with unsupportedError

bishabosha committed:
refactor, error handling in two steps now

bishabosha committed:
err on union in class tparam position

bishabosha committed:
test when ErasedTypeRef is missing from class path

bishabosha committed:
make assertions about metaprogramming features

bishabosha committed:
match types and erased refinements are unsupported

bishabosha committed:
remove comment

bishabosha committed:
avoid resolving ctor of parent

bishabosha committed:
Add mode to track when in an annotation and adjust error message to suit

bishabosha committed:
refactor error handling, test unsupported trees in annotations

bishabosha committed:
Merge readParentFromTerm with readTerm

bishabosha committed:
refactor

bishabosha committed:
only expose TastyName to TreeUnpickler

bishabosha committed:
require TypeName rather than pass boolean flag

bishabosha committed:
reduce dependencies

bishabosha committed:
delete more code

bishabosha committed:
Test open class, dependent methods, trait params

bishabosha committed:
remove CI from PR

bishabosha commented on build.sbt:
```diff
@@ -878,6 +894,75 @@ lazy val junit = project.in(file("test") / "junit")
+lazy val dotc = inputKey[Unit]("compile with dotty")
+lazy val dotcd = inputKey[Unit]("decompile tasty files")
+lazy val scalac = inputKey[Unit]("compile with scalac")
+lazy val runDotty = inputKey[Unit]("run a classfile with the dotty library available")
```
These keys help for testing within the `tasty` project but maybe they are too polluting?

bishabosha replied:
These were removed in https://github.com/scala/scala/pull/8865/commits/d6cca665e19f146ded8d206feac6945fc9090756

bishabosha committed:
Delete SourceFile.scala

was an oversight

bishabosha commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -252,6 +252,7 @@ trait ScalaSettings extends StandardScalaSettings with Warnings {
   val Ymacroexpand    = ChoiceSetting     ("-Ymacro-expand", "policy", "Control expansion of macros, useful for scaladoc and presentation compiler.", List(MacroExpand.Normal, MacroExpand.None, MacroExpand.Discard), MacroExpand.Normal)
   val YmacroFresh     = BooleanSetting    ("-Ymacro-global-fresh-names", "Should fresh names in macros be unique across all compilation units")
   val YmacroAnnotations = BooleanSetting  ("-Ymacro-annotations", "Enable support for macro annotations, formerly in macro paradise.")
+  val YtastyNoAnnotations = BooleanSetting("-Ytasty-no-annotations", "Disable support for reading annotations from TASTy, this will prevent safety features such as pattern match exhaustivity and exhaustion analysis.")
```
```suggestion
  val YtastyNoAnnotations = BooleanSetting("-Ytasty-no-annotations", "Disable support for reading annotations from TASTy, this will prevent safety features such as pattern match exhaustivity and reachability analysis.")
```

bishabosha committed:
support export flag

bishabosha commented on src/compiler/scala/tools/nsc/tasty/TastyReader.scala:
```diff
@@ -0,0 +1,142 @@
+ *  @param end      The position one greater than the last byte to be read
+ *  @param base     The index referenced by the logical zero address Addr(0)
+ */
+class TastyReader(val bytes: Array[Byte], start: Int, end: Int, val base: Int = 0) {
```
These classes are duplicated from the `tasty-core` project in dotty https://github.com/lampepfl/dotty/tree/master/tasty/src/dotty/tools/tasty I hoped we could depend on it, but its compiled with scala 2.13.1, so probably not allowed

bishabosha commented on src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala:
```diff
@@ -0,0 +1,1126 @@
+    nameAtRef: NameRef => TastyName,
+    splices: Seq[Any])(implicit
+    val tasty: Tasty) { self =>
+  import tasty._, FlagSets._
```
main aim here is that `tasty` provides some type aliases and we don't directly select from `tasty.symbolTable`, not  yet converted to abstract types due to directly selecting on symbols here

bishabosha commented on src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:
```diff
@@ -0,0 +1,307 @@
+import scala.tools.nsc.tasty.TastyName
+import scala.tools.nsc.tasty.TastyModes._
+
+trait ContextOps { self: TastyUniverse =>
```
once TreeUnpickler can only see abstract types, then all the `FooOps` traits will still see the concrete types of symbolTable

bishabosha commented on src/compiler/scala/tools/nsc/tasty/bridge/NameOps.scala:
```diff
@@ -0,0 +1,39 @@
+  import self.{symbolTable => u}
+  import TastyName._
+
+  def isConstructorName(name: TastyName) = name == TastyName.Constructor || name == TastyName.MixinConstructor // u.nme.isConstructorName(encodeTastyName(name))
```
```suggestion
  def isConstructorName(name: TastyName) = name == TastyName.Constructor || name == TastyName.MixinConstructor
```

bishabosha commented on src/compiler/scala/tools/nsc/tasty/bridge/TastyCore.scala:
```diff
@@ -0,0 +1,38 @@
+import nsc.symtab, nsc.tasty.TastyUniverse
+
+abstract class TastyCore { self: TastyUniverse =>
+  import self.{symbolTable => u}
```
Ideally this should extend some version where all the types and operations are abstract and thats what TreeUnpickler sees, but perhaps that would be useless, depending on what shared API is extracted for reading TASTy.

bishabosha commented on src/compiler/scala/tools/nsc/tasty/package.scala:
```diff
@@ -0,0 +1,10 @@
+import annotation.unchecked.uncheckedVariance
+
+package object tasty {
+  implicit final class SafeEq[-T](private val t: T @uncheckedVariance) extends AnyVal {
```
assert rhs is subtype of lhs

bishabosha commented on src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala:
```diff
@@ -111,62 +105,12 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
-    *  and a method for re-use elsewhere, because nobody will get this right without
-    *  some higher level facilities.
-    */
-    def extensionMethInfo(extensionMeth: Symbol, origInfo: Type, clazz: Symbol): Type = {
```
I moved this to `internal.Types` as I was not sure how to transform symbols coming from tasty as ExtensionMethods only works on trees, and there is no flag to detect if a symbol of a tree is from tasty

bishabosha committed:
add headers

bishabosha commented on src/compiler/scala/tools/nsc/tasty/ErasedTypeRef.scala:
```diff
@@ -0,0 +1,54 @@
+
+import scala.tools.nsc.tasty.TastyName.{ModuleName, QualifiedName, SimpleName}
+
+case class ErasedTypeRef(arrayDims: Int, qualifiedName: String, isModule: Boolean) {
```
this needs more work to handle names that need encoding

bishabosha replied:
now updated to keep the full name

bishabosha committed:
ensure symbollic classes can be read from erasedtyperef

bishabosha committed:
Simplify refined/rec types

bishabosha committed:
refactor annotations

bishabosha commented on src/library/scala/collection/mutable/StringBuilder.scala:
```diff
@@ -23,7 +23,7 @@ import scala.Predef.{ // unimport char-related implicit conversions to avoid tri
   wrapCharArray => _,
   wrapString => _,
-  //_
+  _
```
```suggestion
 //_
```

bishabosha committed:
refactor TastyFlags

bishabosha committed:
move TastyFlags. Remove Scope dependency

bishabosha committed:
remove TypeBounds dependency

bishabosha committed:
better semantics for symbol.not(flags)

bishabosha commented:
This is ready for review now

bishabosha committed:
better logging

lrytz commented:
> This is ready for review now

Great - it'll take me a week or two to get to, and through it.

bishabosha committed:
Update src/library/scala/collection/mutable/StringBuilder.scala

restore comment

bishabosha commented:
Be sure to try out the flag `-Ydebug-tasty` which gives a lot of information which you can use to trace where in a tasty file you are currently traversing

bishabosha committed:
scan trees does not complete symbols

bishabosha committed:
annotations use lazy val to compute tree

bishabosha committed:
can visit either module or module class first

bishabosha commented:
The last three commits were important bug fixes but I'll stop pushing to this branch now

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1184,11 +1210,65 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
       unpickler.unpickle(bytes, 0, clazz, staticModule, file.name)
+    } else if (isTASTY) {
+
+      def parseTASTYFile(): Array[Byte] = file.underlyingSource match { // TODO: simplify when #3552 is fixed
```
```suggestion
      def parseTASTYFile(): Array[Byte] = file.underlyingSource match { // TODO: simplify when lampepfl/dotty#3552 is fixed
```

in case that comment applies here

lrytz committed:
no more ui.

lrytz committed:
intellij

bishabosha committed:
add CI conditionally

lrytz committed:
fix tasty/test

bishabosha committed:
use commonSettings in tasty subproject

lrytz commented on src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:
```diff
@@ -0,0 +1,407 @@
+          for (decl <- cls.info.decls if decl.isMethod) {
+            if (decl.isParamAccessor) decl.makeNotPrivate(cls)
+            if (!decl.isClassConstructor) {
+              val extensionMeth = decl.newExtensionMethodSymbol(cls.companion, u.NoPosition)
```
just a question: is `cls.companion` guaranteed to be defined?

bishabosha replied:
dotty manually defines `equals$extension` and `hashCode$extension` in the companion so I assume so

bishabosha replied:
Actually, this seems to be false, so I should guard that

bishabosha replied:
actually the companion object is still made even for a basic `class Box(val a: Any) extends AnyVal`, just not the extension methods for equals and hashCode

lrytz reviewed (commented):
some feedback before i forget:

  * Unpickler class names: `ScalacUnpickler` is maybe not the best name - put Tasty in it somewhere? SectionUnpickler, TreeSectionUnpickler: do we need those small wrapper classes?
  * The explanations you wrote in the PR description are really excellent and should make their way into the source code.
  * Maybe you can extend `tastyreader.md` or `tastytest.md` (side-note: merge the two?) with some workflow advice how to effectively work on the tasty rader code.

bishabosha commented:
@lrytz The organisation of `ScalacUnpickler`, `SectionUnpickler` and `TreeSectionUnpickler` are a leftover from the dotty implementation which also has a `CommentUnpickler` and `PositionUnpickler` but they could be merged

bishabosha commented on src/compiler/scala/tools/nsc/tasty/bridge/TypeOps.scala:
```diff
@@ -0,0 +1,464 @@
+      bounds
+  }
+
+  private[bridge] def resolveErasedTypeRef(ref: ErasedTypeRef)(implicit ctx: Context): Type = {
```
this is not correct for converting an ErasedTypeRef of an inner class of a class, in another branch I instead convert Type => ErasedTypeRef

bishabosha committed:
enum values are now case objects

bishabosha committed:
fix bytecode for dotty enum singletons, remove DottyEnumTag constant

bishabosha committed:
refactor erasure for singletons, test more of their behaviour

bishabosha committed:
remove ScalacUnpickler, TreeSectionUnpickler and SectionUnpickler

bishabosha committed:
be more precise with flags from tasty

bishabosha committed:
Remove symbol creation for BIND tags

bishabosha committed:
add documentation

bishabosha committed:
add patmat test

bishabosha committed:
create a single scripted runner for local tasty testing

bishabosha committed:
add AnyVal erasure of enums test

bishabosha committed:
rework documentation

bishabosha committed:
update documentation

bishabosha committed:
eliminate Ident and Annotation from TastyCore

bishabosha committed:
insert parens to nillary applications

bishabosha commented:
This has been rebased for scala 2.13.3

NthPortal commented:
I don't have the compiler knowledge to review this, and I don't have the capacity to obtain the relevant knowledge right now. sorry

bishabosha commented:
@NthPortal sorry, I forced pushed some code and it requested a review

bishabosha committed:
add more refinement tests

lrytz commented on doc/internal/tastyreader.md:
```diff
@@ -0,0 +1,189 @@
+### Entry Points
+
+A classfile is assumed to have an associated TASTy file if it has a `TASTY` annotation (not available through
+Java reflection). This annotation contains a UUID that matches a UUID in the header of a sibling `.tasty` file of the
```
Nit: It's a "classfile attribute", not an annotation. There is confusion about the two terms unfortunately within the Scala compiler (`DeprecatedAttr` in `Definitions` should be `DeprecatedAnnot`), but let's use the right term here...

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1088,19 +1099,26 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
     var innersStart = -1
     var runtimeAnnotStart = -1
+    var TASTYAnnotStart = -1
+    var TASTYAnnotLen = -1
```
Annot/Attr

NthPortal commented:
no worries 💜

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -104,6 +110,11 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
     val symbolTable: ClassfileParser.this.symbolTable.type = ClassfileParser.this.symbolTable
   }
 
+  implicit object TastyUniverse extends TastyUniverse {
```
`implicit` is not needed here

lrytz reviewed (commented):
This looks really great!

As discussed offline, let's try to assign the `children` of sealed types more early. Doing it late (patmat) causes `sym.children` to be incomplete if someone (a macro, compiler plugin) happens to call it early, and it's also a bit ad-hoc.

Ideally it would happen already during tasty-unpickling, similar to what's done in ordinary unpickling. If that fails because completing annotation arguments causes cycles, I can implement what was discussed here: https://github.com/scala/scala/pull/9023#issuecomment-645757880, which we should do anyway.

I don't know why the travis build failed... On jenkins it's green. Do you have an idea?

```
compiling sources in /home/travis/build/scala/scala/test/tasty/pos/src-3 with dotc.
exception occurred while compiling /home/travis/build/scala/scala/test/tasty/pos/src-3/dotty_i3816.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/IO.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Defaults.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Reader.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Functor.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Monad.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Annotated.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/printing/printers.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/rootAnnot.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/HelloWorld.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/DelayedParameterised.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/AnythingIsPossible.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Delayed.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/TraitWithSideEffects.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/symbolicAnnot.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/DelayedInternal.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest_>>>/package.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/InEmpty.scala
java.lang.AssertionError: assertion failed: no extension method found for:
  method ->:[B](y: B): (A, B) with signature Signature(List(1, java.lang.Object),scala.Tuple2) in  <none>
 Candidates:
 
 Candidates (signatures normalized):
  while compiling /home/travis/build/scala/scala/test/tasty/pos/src-3/dotty_i3816.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/IO.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Defaults.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Reader.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Functor.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Monad.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Annotated.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/printing/printers.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/rootAnnot.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/HelloWorld.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/DelayedParameterised.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/AnythingIsPossible.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/Delayed.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/TraitWithSideEffects.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/symbolicAnnot.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest/DelayedInternal.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/tastytest_>>>/package.scala, /home/travis/build/scala/scala/test/tasty/pos/src-3/InEmpty.scala
[error] Test scala.tools.tastytest.TastyTestJUnit.pos failed: java.lang.reflect.InvocationTargetException: null, took 2.296 sec
[error]     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]     at java.lang.reflect.Method.invoke(Method.java:498)
[error]     at scala.tools.tastytest.Dotc$.$anonfun$processMethod$3(Dotc.scala:19)
[error]     at scala.tools.tastytest.Dotc$.$anonfun$processMethod$2(Dotc.scala:18)
[error]     at scala.tools.tastytest.Dotc$.dotc(Dotc.scala:38)
[error]     at scala.tools.tastytest.TastyTest$.dotcPos(TastyTest.scala:145)
[error]     at scala.tools.tastytest.TastyTest$.$anonfun$posSuite$8(TastyTest.scala:141)
[error]     at scala.util.Success.flatMap(Try.scala:258)
[error]     at scala.tools.tastytest.TastyTest$.$anonfun$posSuite$7(TastyTest.scala:32)
[error]     at scala.util.Success.flatMap(Try.scala:258)
[error]     at scala.tools.tastytest.TastyTest$.$anonfun$posSuite$4(TastyTest.scala:31)
[error]     at scala.util.Success.flatMap(Try.scala:258)
[error]     at scala.tools.tastytest.TastyTest$.posSuite(TastyTest.scala:29)
[error]     at scala.tools.tastytest.TastyTestJUnit.pos(TastyTestJUnit.scala:23)
[error]     ...
[error] Caused by: java.lang.AssertionError: assertion failed: no extension method found for:
[error] 
[error]   method ->:[B](y: B): (A, B) with signature Signature(List(1, java.lang.Object),scala.Tuple2) in  <none>
[error] 
[error]  Candidates:
[error] 
[error]  
[error] 
[error]  Candidates (signatures normalized):
[error] 
[error]  
[error]     at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
[error]     at dotty.tools.dotc.transform.ExtensionMethods$.extensionMethod$$anonfun$1(ExtensionMethods.scala:195)
[error]     at dotty.tools.dotc.core.Phases.atPhase$$anonfun$1(Phases.scala:35)
[error]     at dotty.tools.dotc.core.Periods.atPhase(Periods.scala:25)
[error]     at dotty.tools.dotc.core.Phases.atPhase(Phases.scala:35)
[error]     at dotty.tools.dotc.core.Contexts$Context.atPhase(Contexts.scala:75)
[error]     at dotty.tools.dotc.transform.ExtensionMethods$.extensionMethod(ExtensionMethods.scala:208)
[error]     at dotty.tools.dotc.transform.VCInlineMethods.rewire(VCInlineMethods.scala:69)
[error]     at dotty.tools.dotc.transform.VCInlineMethods.rewireIfNeeded(VCInlineMethods.scala:96)
[error]     at dotty.tools.dotc.transform.VCInlineMethods.transformApply(VCInlineMethods.scala:106)
[error]     at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:611)
[error]     at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:612)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:272)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTrees$$anonfun$1(MegaPhase.scala:420)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:420)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:328)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTrees$$anonfun$1(MegaPhase.scala:420)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTrees(MegaPhase.scala:420)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:271)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.mapValDef$1(MegaPhase.scala:234)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:238)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:404)
[error]     at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:409)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:339)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:404)
[error]     at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:409)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:286)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:245)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:248)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:404)
[error]     at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:409)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:339)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStat$2(MegaPhase.scala:404)
[error]     at dotty.tools.dotc.transform.MegaPhase.$anonfun$1(MegaPhase.scala:409)
[error]     at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
[error]     at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:356)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:359)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
[error]     at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:415)
[error]     at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:427)
[error]     at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:318)
[error]     at scala.collection.immutable.List.map(List.scala:250)
[error]     at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:319)
[error]     at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:167)
[error]     at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]     at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]     at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]     at dotty.tools.dotc.Run.runPhases$5(Run.scala:177)
[error]     at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:185)
[error]     at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error]     at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
[error]     at dotty.tools.dotc.Run.compileUnits(Run.scala:192)
[error]     at dotty.tools.dotc.Run.compileSources(Run.scala:129)
[error]     at dotty.tools.dotc.Run.compile(Run.scala:112)
[error]     at dotty.tools.dotc.Driver.doCompile(Driver.scala:36)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:189)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:158)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:170)
[error]     at dotty.tools.dotc.Main.process(Main.scala)
[error]     ... 49 more
```

bishabosha committed:
TASTY is classfile attribute

bishabosha committed:
Force annotations in TreeUnpickler and not Pattern Matching

This commit is a squash of several commits to atomically add the feature:

When indexing a scope for the first time to set up completers, add
lazy annotations to symbols. At the end of the indexing, force the annotations
and if one is scala.annotation.internal.Child[T], add the symbol of T as a
sealed child. If we are not indexing statements, such as completing the
a method and reading its parameters, force annotations as soon as each symbol
is initialised, this is ok because the owner tree for the definition is also
initialised so forward references can be safely accessed.

force sealed children after indexing scope

cache RepeatedAnnot

force all annotations early, catching more errors

Also force annotations for lookahead contexts.

It would be good to see if perhaps we can share a single annotation cache per
tasty file and force from one place.

Force annotations for each new symbol
Additionally, when sequencing statements, only force after the whole sequence is indexed.
Forceable annotations are now stored in and forced from the initial context.
Made it more lazy to print parents of a class.
If a LambdaPolyType is in a New tree, unwrap it.

force annotations at end of createMemberSymbol

only make lazy annotation if we delay its evaluation

bishabosha commented:
so I now force the sealed children annotations as soon as all definitions are seen in a scope

bishabosha commented on src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala:
```diff
@@ -0,0 +1,1115 @@
+            skipTree()
+        }
+      }
+      ctx.forceAnnotations()
```
in the present implementation this actually does not deal with the case where a lookahead symbol definition also declares annotations, e.g.
```scala
object DelayedChained {

  sealed trait Root

  final class LeafA extends Root

  object Internal {

    sealed trait Branch extends Root

    object Deeper {
      final class LeafB extends Branch
      final class LeafC extends Branch
    }

  }

}
```

bishabosha replied:
new commit adds so that when we lookahead to create a symbol we also force annotations collected there.

bishabosha commented:
forcing all annotations early is causing some cyclic referencing which needs to be looked at, such as this code
```scala
object DefAnnots {

  def withAnnotatedAnnot(@Wrapper.annotatedAnnot arg: Any): Any = arg

  object Wrapper {
    class annot extends StaticAnnotation
    @annot class annotatedAnnot extends StaticAnnotation
  }

}
```

bishabosha committed:
Delay forcing annotations in template body until parents are added to class info.

Before this, inherited members may not be visible at the point of forcing the annotation
on a definition in the template.

Also moved symbol cache to the initial context so that child contexts all pass to the same
cache.

bishabosha committed:
Find overloads by converting types to ErasedTypeRef.

Inner classes can not be found with Mirror getRequiredClass, so
we should have another way to compare types to signatures, this
proposes to instead convert method signatures to
Signature[ErasedTypeRef]

bishabosha committed:
Merge pull request #106 from scalacenter/tasty/pick-erased-type-ref

Read inner classes in signatures, force annotations after template parents

lrytz committed:
Fix JavaUniverseForce

bishabosha committed:
Merge pull request #107 from lrytz/pr8865

Fix JavaUniverseForce
